File: negotiation-cordapp/contracts-java/src/test/java/negotiation/contracts/AcceptanceContractTests.java
package negotiation.contracts;

import com.google.common.collect.ImmutableList;
import negotiation.states.ProposalState;
import negotiation.states.TradeState;
import net.corda.testing.core.DummyCommandData;
import net.corda.testing.core.TestIdentity;
import net.corda.testing.node.MockServices;
import org.junit.Test;

import java.time.Instant;

import static net.corda.testing.node.NodeTestUtils.ledger;

public class AcceptanceContractTests {

    private MockServices ledgerServices = new MockServices(ImmutableList.of("negotiation.contracts"));
    private TestIdentity alice = new TestIdentity(new net.corda.core.identity.CordaX500Name("alice","New York", "US"));
    private TestIdentity bob = new TestIdentity(new net.corda.core.identity.CordaX500Name("bob","Tokyo", "JP"));
    private TestIdentity charlie = new TestIdentity(new net.corda.core.identity.CordaX500Name("charlie","London", "GB"));

    @Test
    public void proposalAcceptanceTransactionsHaveOnlyOneInputAndOneOutputState(){
        ledger(ledgerServices, ledger ->{
            ledger.transaction(tx ->{
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Accept());
                tx.tweak(tdsl->{
                    tdsl.input(ProposalAndTradeContract.ID, new ProposalState(1,alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                    tdsl.input(ProposalAndTradeContract.ID, new ProposalState(1,alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                    tdsl.fails();
                    return null;
                });
                tx.tweak(tdsl -> {
                    tdsl.output(ProposalAndTradeContract.ID, new TradeState(1, alice.getParty(), bob.getParty()));
                    tdsl.output(ProposalAndTradeContract.ID, new TradeState(1, alice.getParty(), bob.getParty()));
                    tdsl.fails();
                    return null;
                });
                tx.input(ProposalAndTradeContract.ID,new ProposalState(1, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                tx.output(ProposalAndTradeContract.ID, new TradeState(1,alice.getParty(), bob.getParty()));
                tx.verifies();
                return null;
            });
            return null;
        });
    }

    @Test
    public void proposalAcceptanceTransactionsHaveInputOfTypeProposalStateAndOutputOfTypeTradeState(){
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Accept());
                tx.tweak(tdsl ->{
                    tdsl.input(ProposalAndTradeContract.ID,new TradeState(1,alice.getParty(), bob.getParty()));
                    tdsl.fails();
                    return null;
                });
                tx.tweak(tdsl ->{
                    tdsl.output(ProposalAndTradeContract.ID, new ProposalState(1, alice.getParty(),bob.getParty(), alice.getParty(), bob.getParty()));
                    tdsl.fails();
                    return null;
                });

                tx.input(ProposalAndTradeContract.ID, new ProposalState(1, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                tx.output(ProposalAndTradeContract.ID, new TradeState(1, alice.getParty(), bob.getParty()));
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    @Test
    public void proposalAcceptanceTransactionsHaveExactlyOneCommandOfTypeAccept(){
        ledger(ledgerServices, ledger ->{
            ledger.transaction(tx ->{
                tx.input(ProposalAndTradeContract.ID, new ProposalState(1, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                tx.output(ProposalAndTradeContract.ID, new TradeState(1, alice.getParty(), bob.getParty()));
                tx.tweak(tdsl -> {
                    tdsl.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()),  DummyCommandData.INSTANCE);
                    tdsl.fails();
                    return null;
                });
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Accept());
                tx.verifies();
                return null;
            });
            return null;
        });
    }

    @Test
    public void inputProposalStateAndOutputTradeStateShouldHaveExactlyTheSameAmount(){
        ledger(ledgerServices, ledger ->{
            ledger.transaction(tx ->{
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Accept());
                tx.tweak(tdsl ->{
                    tdsl.input(ProposalAndTradeContract.ID,new ProposalState(1, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                    tdsl.output(ProposalAndTradeContract.ID, new TradeState(2, alice.getParty(), bob.getParty()));
                    tdsl.fails();
                    return null;
                });
                tx.tweak(tdsl ->{
                    tdsl.input(ProposalAndTradeContract.ID, new ProposalState(2, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                    tdsl.output(ProposalAndTradeContract.ID, new TradeState(1, alice.getParty(), bob.getParty()));
                    tdsl.fails();
                    return null;
                });
                tx.input(ProposalAndTradeContract.ID,new ProposalState(1, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                tx.output(ProposalAndTradeContract.ID, new TradeState(1, alice.getParty(), bob.getParty()));
                tx.verifies();
                return null;
            });
            return null;
        });
    }


    @Test
    public void buyerAndSellerAreUnmodifiedInTheOutput(){
        ledger(ledgerServices, ledger ->{
            ledger.transaction(tx ->{
                tx.input(ProposalAndTradeContract.ID, new ProposalState(1, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                tx.output(ProposalAndTradeContract.ID, new TradeState(1, alice.getParty(), bob.getParty()));
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Accept());
                tx.tweak(tdsl ->{
                    tdsl.output(ProposalAndTradeContract.ID, new TradeState(1, alice.getParty(),  charlie.getParty()));
                    tdsl.fails();
                    return null;
                });
                tx.tweak(tdsl ->{
                    tdsl.output(ProposalAndTradeContract.ID, new TradeState(1, alice.getParty(),  charlie.getParty()));
                    tdsl.fails();
                    return null;
                });
                tx.verifies();
                return null;
            });
            return null;
        });
    }

    @Test
    public void proposalAcceptanceTransactionsHaveTwoRequiredSignersTheProposerAndProposee(){
        ledger(ledgerServices, ledger ->{
            ledger.transaction(tx ->{
                tx.input(ProposalAndTradeContract.ID, new ProposalState(1, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                tx.output(ProposalAndTradeContract.ID, new TradeState(1, alice.getParty(), bob.getParty()));
                tx.tweak(tdsl ->{
                    tdsl.command(ImmutableList.of(alice.getPublicKey(), charlie.getPublicKey()), new ProposalAndTradeContract.Commands.Accept());
                    tdsl.fails();
                    return null;
                });
                tx.tweak(tdsl -> {
                    tdsl.command(ImmutableList.of(charlie.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Accept());
                    tdsl.fails();
                    return null;
                });
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Accept());
                tx.verifies();
                return null;
            });
            return null;
        });
    }

    @Test
    public void proposalAcceptanceTransactionsHaveNoTimeStamp(){
        ledger(ledgerServices, ledger ->{
            ledger.transaction(tx ->{
                tx.input(ProposalAndTradeContract.ID, new ProposalState(1, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                tx.output(ProposalAndTradeContract.ID, new TradeState(1, alice.getParty(), bob.getParty()));
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Accept());
                tx.tweak(tdsl ->{
                    tdsl.timeWindow(Instant.now());
                    tdsl.fails();
                    return null;
                });
                return null;
            });
            return null;
        });
    }
}

File: negotiation-cordapp/contracts-kotlin/src/test/kotlin/negotiation/contracts/AcceptanceContractTests.kt
package negotiation.contract

import negotiation.contracts.ProposalAndTradeContract
import negotiation.states.ProposalState
import negotiation.states.TradeState
import net.corda.core.identity.CordaX500Name
import net.corda.testing.core.DummyCommandData
import net.corda.testing.core.TestIdentity
import net.corda.testing.node.MockServices
import net.corda.testing.node.ledger
import org.junit.Test
import java.time.Instant

class AcceptanceContractTests {
    private val ledgerServices = MockServices(listOf("negotiation.contracts"))
    private val alice = TestIdentity(CordaX500Name("alice", "New York", "US"))
    private val bob = TestIdentity(CordaX500Name("bob", "Tokyo", "JP"))
    private val charlie = TestIdentity(CordaX500Name("charlie", "London", "GB"))

    @Test
    fun `proposal acceptance transactions have only one input and one output state`() {
        ledgerServices.ledger{
            transaction {

                command(listOf(alice.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Accept())
                tweak {
                    input(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                    input(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                    fails()
                }
                tweak {

                    output(ProposalAndTradeContract.ID, TradeState(1, alice.party, bob.party))
                    output(ProposalAndTradeContract.ID, TradeState(1, alice.party, bob.party))
                    fails()
                }
                input(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                output(ProposalAndTradeContract.ID, TradeState(1, alice.party, bob.party))
                verifies()
            }
        }
    }

    @Test
    fun `proposal acceptance transactions have input of type Proposal State and output of type TradeState`() {
        ledgerServices.ledger{
            transaction {
                command(listOf(alice.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Accept())
                tweak {
                    input(ProposalAndTradeContract.ID, TradeState(1, alice.party, bob.party))
                    fails()
                }
                tweak {
                    output(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                    fails()
                }
                input(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                output(ProposalAndTradeContract.ID, TradeState(1, alice.party, bob.party))
                verifies()
            }
        }
    }

    @Test
    fun `proposal acceptance transactions have exactly one command of type Accept`() {
        ledgerServices.ledger {
            transaction {
                input(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                output(ProposalAndTradeContract.ID, TradeState(1, alice.party, bob.party))
                tweak {
                    command(listOf(alice.publicKey, bob.publicKey), DummyCommandData)
                    fails()
                }
                command(listOf(alice.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Accept())
                verifies()
            }
        }
    }

    @Test
    fun `input proposal state and output trade state should have exactly same amounts`() {
        ledgerServices.ledger{
            transaction {
                command(listOf(alice.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Accept())
                tweak {
                    input(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                    output(ProposalAndTradeContract.ID, TradeState(2, alice.party, bob.party))
                    fails()
                }
                tweak {
                    input(ProposalAndTradeContract.ID, ProposalState(2, alice.party, bob.party, alice.party, bob.party))
                    output(ProposalAndTradeContract.ID, TradeState(1, alice.party, bob.party))
                    fails()
                }
                input(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                output(ProposalAndTradeContract.ID, TradeState(1, alice.party, bob.party))
                verifies()
            }
        }
    }

    @Test
    fun `buyer and seller are unmodified in the output`() {
        ledgerServices.ledger {
            transaction {
                input(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                output(ProposalAndTradeContract.ID, TradeState(1, alice.party, bob.party))
                command(listOf(alice.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Accept())
                tweak {
                    output(ProposalAndTradeContract.ID, TradeState(1, alice.party, charlie.party))
                    fails()
                }
                tweak {
                    output(ProposalAndTradeContract.ID, TradeState(1, charlie.party, bob.party))
                    fails()
                }
                tweak {
                    output(ProposalAndTradeContract.ID, TradeState(1, bob.party, bob.party))
                    fails()
                }
                verifies()
            }
        }
    }

    @Test
    fun `proposal acceptance transactions have two required signers - the proposer and the proposee`() {
        ledgerServices.ledger {
            transaction {
                input(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                output(ProposalAndTradeContract.ID, TradeState(1, alice.party, bob.party))
                tweak {
                    command(listOf(alice.publicKey, charlie.publicKey), ProposalAndTradeContract.Commands.Accept())
                    fails()
                }
                tweak {
                    command(listOf(charlie.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Accept())
                    fails()
                }
                command(listOf(alice.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Accept())
                verifies()
            }
        }
    }

    @Test
    fun `proposal acceptance transactions have no timestamp`() {
        ledgerServices.ledger {
            transaction {
                input(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                output(ProposalAndTradeContract.ID, TradeState(1, alice.party, bob.party))
                command(listOf(alice.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Accept())
                tweak {
                    timeWindow(Instant.now())
                    fails()
                }
            }
        }
    }
}